functionary/train/train_lora.py

```diff
@@ -0,0 +1,396 @@
+    ddp = world_size != 1
+    if lora_args.q_lora:
+        if ddp and training_args.fsdp:
+            print("FSDP is incompatible with QLORA")
+        device_map = {"": int(os.environ.get("LOCAL_RANK") or 0)} if ddp else None
+        if len(training_args.fsdp) > 0 or deepspeed.is_deepspeed_zero3_enabled():
+            print("FSDP and ZeRO3 are both currently incompatible with QLoRA.")
+    return device_map
+
+
+def load_model_with_rope_scaling(
+    model_args: ModelArguments, training_args: TrainingArguments, lora_args: LoraArguments
+) -> transformers.AutoModelForCausalLM:
+    config = transformers.AutoConfig.from_pretrained(
+        model_args.model_name_or_path,
+        cache_dir=training_args.cache_dir,
+    )
+    orig_ctx_len = getattr(config, "max_position_embeddings", None)
+    if orig_ctx_len and training_args.model_max_length > orig_ctx_len:
+        scaling_factor = float(math.ceil(training_args.model_max_length / orig_ctx_len))
+        config.rope_scaling = {"type": "linear", "factor": scaling_factor}
+    config.use_cache = False
+
+    compute_dtype = torch.float16 if training_args.fp16 else (torch.bfloat16 if training_args.bf16 else torch.float32)
+
+    model = transformers.AutoModelForCausalLM.from_pretrained(
+        model_args.model_name_or_path,
+        config=config,
+        cache_dir=training_args.cache_dir,
+        device_map=get_device_map(training_args, lora_args),
+        quantization_config=BitsAndBytesConfig(
+            load_in_4bit=True,
+            bnb_4bit_use_double_quant=True,
+            bnb_4bit_quant_type="nf4",
+            use_flash_attention_2=True,
+            bnb_4bit_compute_dtype=compute_dtype,
+        )
+        if lora_args.q_lora
+        else None,
+    )
+    return model
+
+
+def prepare_model_for_training(
+    model: transformers.AutoModelForCausalLM, training_args: TrainingArguments, lora_args: LoraArguments
+):
+    if lora_args.lora_target_modules == "all":
+        target_modules = find_all_linear_names(model)
+    else:
+        modules = lora_args.lora_target_modules.split(" ")
+        target_modules = [mod.strip() for mod in modules if len(mod.strip()) > 0]
+
+    print_rank0("target_modules: ", target_modules)
+    lora_config = LoraConfig(
+        r=lora_args.lora_r,
+        lora_alpha=lora_args.lora_alpha,
+        target_modules=target_modules,
+        lora_dropout=lora_args.lora_dropout,
+        bias=lora_args.lora_bias,
+        task_type="CAUSAL_LM",
+        modules_to_save=["lm_head", "embed_tokens"],  # because we retrain the embedding
+    )
+
+    if lora_args.q_lora:
+        model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=training_args.gradient_checkpointing)
+
+    model = get_peft_model(model, lora_config)
+
+    if training_args.gradient_checkpointing:
+        model.enable_input_require_grads()
+
+    model.config.use_cache = False
+    print_trainable_parameters(model)
+    return model
+
+
+def maybe_zero_3(param):
```

Comment:
can you add docstring or comments what this does

Reply:
Actually I copy this from train_lora fastchat, I will make a comment there
[maybe_zero_3](https://github.com/lm-sys/FastChat/blob/main/fastchat/train/train_lora.py#L68)